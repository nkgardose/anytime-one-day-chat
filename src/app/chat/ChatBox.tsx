/** @jsxImportSource @emotion/react */
import { useLazyQuery, useMutation, useQuery } from '@apollo/client'
import localforage from 'localforage'
import { useEffect, useMemo, useRef, useState } from 'react'
import { useParams } from 'react-router-dom'
import MESSAGE_POST from '../../apollo/mutation/messagePost'
import FETCH_LATEST_MESSAGE from '../../apollo/query/fetchLatestMessages'
import FETCH_MORE_MESSAGES from '../../apollo/query/fetchMoreMessages'
import Header from '../../components/header'
import useStore from '../../hooks/useStore'
import { MessagesAction, MessageStatus, type Message } from '../../reducers/Messages'
import MessageInput, { type IMessageInputHandle } from './MessageInput'
import Messages from './Messages'
import { chatBox } from './style'

const ChatBox = (): JSX.Element => {
  const { store, dispatch } = useStore()
  const { channel } = useParams()
  const { data, loading } = useQuery(FETCH_LATEST_MESSAGE, {
    variables: { channelId: channel ?? '' },
    fetchPolicy: 'network-only'
  })
  const [postMessage] = useMutation(MESSAGE_POST)
  const messageInput = useRef<IMessageInputHandle>(null)
  const [message, setMessage] = useState('')

  const messages = useMemo(() => {
    if (store.messages !== undefined && channel !== undefined) return store.messages[channel] ?? []
    return []
  }, [store.messages, channel])
  const [fetchMoreMessage] = useLazyQuery(FETCH_MORE_MESSAGES, {
    variables: { channelId: channel ?? '', messageId: messages.length > 0 ? messages[0].messageId : '', old: true }
  })

  useEffect(() => {
    if (data?.fetchLatestMessages !== undefined && channel !== undefined) {
      dispatch({
        type: MessagesAction.SET_MESSAGES,
        messages: data.fetchLatestMessages
          .map(
            (message: any): Message => ({
              messageId: message.messageId,
              text: message.text,
              timestamp: new Date(message.datetime),
              userId: message.userId,
              status: MessageStatus.SENT
            })
          )
          .sort((a: Message, b: Message) => a.timestamp.getTime() - b.timestamp.getTime()),
        channel
      })
    }
  }, [data])

  useEffect(() => {
    if (channel !== undefined)
      void localforage.getItem(`${channel}.message`).then((message = '') => {
        messageInput.current?.setValue(message as string)
      })
  }, [channel])

  useEffect(() => {
    if (channel !== undefined) void localforage.setItem(`${channel}.message`, message)
  }, [message])

  const onChange = (value: string): void => {
    setMessage(value)
  }

  const sendMessage = async (user: string, channel: string): Promise<void> => {
    try {
      const res = await postMessage({
        variables: {
          channelId: channel,
          text: message,
          userId: user
        }
      })

      if (res.data !== undefined) {
        const message = res.data.postMessage
        dispatch({
          type: MessagesAction.SEND_MESSAGE,
          channel,
          message: {
            messageId: message.messageId,
            text: message.text,
            timestamp: new Date(message.datetime),
            userId: user,
            status: MessageStatus.SENT
          }
        })
      }
      if (res.errors !== undefined) {
        dispatch({
          type: MessagesAction.SEND_MESSAGE,
          channel,
          message: {
            messageId: Math.random().toString(),
            text: message,
            timestamp: new Date(),
            userId: user,
            status: MessageStatus.ERROR
          }
        })
      }
    } catch (e) {
      dispatch({
        type: MessagesAction.SEND_MESSAGE,
        channel,
        message: {
          messageId: Math.random().toString(),
          text: message,
          timestamp: new Date(),
          userId: user,
          status: MessageStatus.ERROR
        }
      })
    }
  }

  const onSend = (): void => {
    if (channel !== undefined && store.auth?.activeUser !== undefined) {
      messageInput.current?.clear()
      if (store.auth.activeUser !== undefined && channel !== undefined) {
        void sendMessage(store.auth.activeUser, channel)
      }
    }
  }

  const onFetchOld = async (): Promise<void> => {
    try {
      const { data } = await fetchMoreMessage()
      if (data !== undefined && channel !== undefined) {
        dispatch({
          type: MessagesAction.SET_MESSAGES,
          messages: [
            ...messages,
            ...(data.fetchMoreMessages ?? []).map(
              (message: any): Message => ({
                messageId: message.messageId,
                text: message.text,
                timestamp: new Date(message.datetime),
                userId: message.userId,
                status: MessageStatus.SENT
              })
            )
          ],
          channel
        })
      }
    } catch (e) {
      console.warn(e)
    }
  }

  return (
    <div css={chatBox}>
      <Header>{channel} Channel</Header>
      <Messages messages={messages ?? []} onFetchOld={onFetchOld} />
      <MessageInput ref={messageInput} loading={loading} onChange={onChange} sendMessage={onSend} />
    </div>
  )
}

export default ChatBox
