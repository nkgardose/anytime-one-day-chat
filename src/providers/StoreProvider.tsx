import React, { createContext, useMemo, useReducer } from 'react'
import AuthReducer, {
  type AuthActionTypes,
  type AuthState
} from '../reducers/Auth'
import { users } from '../utils/constants'

/**
 * Reducers
 */

type Reducers = React.Reducer<AuthState, AuthActionTypes>

/**
 * Dispatch & Action
 */

type Action = AuthActionTypes

type StoreDispatch = React.Dispatch<Action>

/**
 * Store State
 */

interface Store extends Record<string, any> {
  auth?: AuthState
}

/**
 * Context & Providers
 */

export interface IStoreContext {
  store: Store
  dispatch: StoreDispatch
}

interface IStoreProvider extends React.PropsWithChildren {
  defaultStore?: Store
}

const initialState: Store = {
  auth: {
    users,
    activeUser: users[0]
  }
}

export const StoreContext = createContext<IStoreContext>({
  store: initialState,
  dispatch: () => {}
})

const combineReducer =
  (reducers: Record<keyof Store, Reducers>) => (store: Store, action: Action) =>
    Object.keys(reducers).reduce(
      (accumulatedStore, currentReducerKey) => ({
        ...accumulatedStore,
        [currentReducerKey]: reducers[currentReducerKey](
          accumulatedStore[currentReducerKey],
          action
        )
      }),
      store
    )

const StoreProvider: React.FunctionComponent<IStoreProvider> = ({
  defaultStore,
  children
}) => {
  const [state, dispatch] = useReducer(
    combineReducer({ auth: AuthReducer }),
    defaultStore ?? initialState
  )

  const store = useMemo(() => ({ state, dispatch }), [state, dispatch])

  return (
    <StoreContext.Provider
      value={{ store: store.state, dispatch: store.dispatch }}
    >
      {children}
    </StoreContext.Provider>
  )
}

export default StoreProvider
