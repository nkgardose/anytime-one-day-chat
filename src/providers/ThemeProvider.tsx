import React from 'react'
import { ThemeProvider as Theme } from '@emotion/react'
import theme from '../style/theme'

const ThemeProvider: React.FunctionComponent<React.PropsWithChildren> = ({
  children
}) => <Theme theme={theme}>{children}</Theme>

export default ThemeProvider
