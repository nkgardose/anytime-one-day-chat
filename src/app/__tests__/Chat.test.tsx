import '@testing-library/jest-dom'
import { act, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { createMemoryRouter, RouterProvider } from 'react-router-dom'
import Providers from '../../providers'
import StoreProvider from '../../providers/StoreProvider'
import ThemeProvider from '../../providers/ThemeProvider'
import ChatBox from '../chat'

describe('Channel', () => {
  beforeEach(() => {
    window.HTMLElement.prototype.scrollIntoView = function () {}
  })
  it('renders properly', () => {
    const { container } = render(
      <Providers>
        <ChatBox />
      </Providers>
    )

    expect(container).toMatchSnapshot()
  })
  it('renders correct content', async () => {
    await act(async () => {
      render(
        <ThemeProvider>
          <StoreProvider>
            <RouterProvider
              router={createMemoryRouter(
                [
                  {
                    path: '/chat/:channel',
                    element: <ChatBox />
                  }
                ],
                {
                  initialEntries: ['/chat/General'],
                  initialIndex: 0
                }
              )}
            />
          </StoreProvider>
        </ThemeProvider>
      )
    })

    expect(screen.getByText('General Channel')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Type a message...')).toBeInTheDocument()
  })
  it('updates the contentEditable and sends the message', async () => {
    await act(async () => {
      render(
        <ThemeProvider>
          <StoreProvider>
            <RouterProvider
              router={createMemoryRouter(
                [
                  {
                    path: '/chat/:channel',
                    element: <ChatBox />
                  }
                ],
                {
                  initialEntries: ['/chat/General'],
                  initialIndex: 0
                }
              )}
            />
          </StoreProvider>
        </ThemeProvider>
      )
    })
    const input = screen.getByPlaceholderText('Type a message...')

    userEvent.click(input)
    await act(async () => {
      userEvent.keyboard('ABC')
    })

    expect(input.textContent).toBe('ABC')

    await act(async () => {
      userEvent.click(screen.getByLabelText('send'))
    })

    expect(input.textContent).toBe('')

    userEvent.click(input)
    await act(async () => {
      userEvent.keyboard('A B C')
    })

    await act(async () => {
      userEvent.type(input, '{enter}')
    })

    expect(input.textContent).toBe('')

    expect(screen.getByText('ABC')).toBeInTheDocument()
    expect(screen.getByText('A B C')).toBeInTheDocument()
    expect(screen.getAllByText('Joyse').length).toBe(1)
  })
})
