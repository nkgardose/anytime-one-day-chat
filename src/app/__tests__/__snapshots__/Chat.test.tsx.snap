// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel renders properly 1`] = `
<div>
  <div
    class="css-yu2ud0-ChatBox"
  >
    <nav
      class="css-17wuva9-Header"
    >
       Channel
    </nav>
    <div
      class="css-zp5xvb-Messages"
    />
    <div
      class="css-1e2k6mq-MessageInput"
    >
      <div
        class="css-xqsv07-MessageInput"
        contenteditable="true"
        placeholder="Type a message..."
        role="textbox"
        tabindex="0"
      />
      <i
        aria-label="send"
        class="css-186inh8-MessageInput"
        role="button"
        tabindex="0"
      >
        <svg
          fill="currentColor"
          height="25"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
          />
        </svg>
      </i>
    </div>
  </div>
</div>
`;
